### PR TITLE
Avoid deadlock when restarting SIP UDP transport due to holding pjsua lock

### PR DESCRIPTION
**Pull request by trengginas**
This ticket is to address #2717 by not holding PJSUA lock when calling `pjsip_udp_transport_restart2()`.

### CODE REVIEW

**Comment by nanangizz** [comment]
Theoretically PJSUA lock must be held while accessing any PJSUA state (e.g: `pjsua_var`), so better try to find another approach find, e.g: unlock temporarily in the loop spin.

File: pjsip/src/pjsua-lib/pjsua_core.c
Comment (nanangizz) [inline pjsip/src/pjsua-lib/pjsua_core.c:3751]: Why introduce new APIs, instead of simply unlocking PJSUA lock before invoking `pjsip_udp_transport_restart*()` and re-locking later after?

**Comment by nanangizz** [comment]
Hmm.. I didn't analyze the problem in detail, so I don't really know whether callbacks in the previous patch are really needed or not. Also no idea whether the simpler approach is better.

So, could you describe the previous approach further. E.g: Is there any mutexes other than PJSUA, perhaps for other apps? And does it need to be unlocked right before the spin loop (as opposed to before `pjsip_udp_transport_restart()`)?

**Comment by trengginas** [comment]
The deadlock issue is because prior to calling `pjsip_udp_transport_restart2()`, `pjsua_handle_ip_change()` will acquire pjsua lock.
When restarting UDP transport, it will wait for the read spin loop to finish. This will trigger a deadlock if there's a new incoming message which in turn tried to acquire the pjsua lock.
The alternatives:
- Make sure that there's no locks held before calling `pjsip_udp_transport_restart2()`
- Provide a mechanism to release the lock before waiting for the read spin loop to finish.

The new API on the previous alternative will provide mechanism for user to be able to release the lock before waiting for the read loop spin to finish.
However, since currently only `pjsua_handle_ip_change()` that will call `pjsip_udp_transport_restart2()`, releasing the pjsua lock before restarting the transport should be simpler and provide a solution to #2717.

**Comment by nanangizz** [comment]
> The deadlock issue is because prior to calling `pjsip_udp_transport_restart2()`, `pjsua_handle_ip_change()` will acquire pjsua lock.
> When restarting UDP transport, it will wait for the read spin loop to finish. This will trigger a deadlock if there's a new incoming message which in turn tried to acquire the pjsua lock.
> The alternatives:
> 
> * Make sure that there's no locks held before calling `pjsip_udp_transport_restart2()`
> * Provide a mechanism to release the lock before waiting for the read spin loop to finish.
> 
> The new API on the previous alternative will provide mechanism for user to be able to release the lock before waiting for the read loop spin to finish.
> However, since currently only `pjsua_handle_ip_change()` that will call `pjsip_udp_transport_restart2()`, releasing the pjsua lock before restarting the transport should be simpler and provide a solution to #2717.

As `pjsip_udp_transport_restart2()` is public API, I don't think it is safe to assume that `pjsua_handle_ip_change()` is the only user. So, the question is: can we just suggest app (i.e: via specs docs) to release any locks before invoking `pjsip_udp_transport_restart2()`? Or it should be in callback form for some reason (please explain if it should)?

**Comment by sauwming** [comment]
Since there's already a delay restart mechanism in `restart_listener()`, perhaps we can just use this?
i.e. if there's read spin loop, return PJ_EBUSY and try restarting it again later.

**Comment by trengginas** [comment]
Personally, I think adding a warning in the doc should be sufficient.

**Comment by trengginas** [comment]
If we return to the read spin loop, then there's a chance that the restart
will need to be retried indefinitely provided incoming messages are still
going on.

**Comment by nanangizz** [comment]
> Personally, I think adding a warning in the doc should be sufficient.

Okay, so just need to update the `pjsip_udp_transport_restart*()` docs.

**Comment by nanangizz** [comment]
Just did a quick check in `pjsip_udp_transport_restart2()` there is also a call to `pj_ioqueue_unregister()`, can this cause deadlock too, e.g: wait for ioqueue key lock being held by the read op?

**Comment by trengginas** [comment]
It shouldn't cause deadlock for `pjsip_udp_transport_restart*()`, since any locks acquired in `pj_ioqueue_unregister()` will be released once the read loop wait is executed. 
Or perhaps you meant deadlock inside `pj_ioqueue_unregister()` ?

**Comment by nanangizz** [comment]
Ah right, as now any mutex should be released by app prior to invoking `pjsip_udp_transport_restart*()`, `pj_ioqueue_unregister()` eventually will be executed.

**Review by nanangizz — APPROVED** [review]
It'd be better to wait for confirmation from reporter before merging this.

**Review by sauwming — APPROVED** [review]

**Comment by dstd** [comment]
yep, it looks fixed now – deadlocks didn't appear anymore

**Pull request merged**
